providers/fips/fipsprov.c

```diff
@@ -88,12 +139,71 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
         case OSSL_FUNC_CORE_GET_PARAMS:
             c_get_params = OSSL_get_core_get_params(in);
             break;
+        case OSSL_FUNC_CORE_PUT_ERROR:
+            c_put_error = OSSL_get_core_put_error(in);
+            break;
+        case OSSL_FUNC_CORE_ADD_ERROR_VDATA:
+            c_add_error_vdata = OSSL_get_core_add_error_vdata(in);
+            break;
         /* Just ignore anything we don't understand */
         default:
             break;
         }
     }
 
+    if (!ex_data_init_intern())
+        return 0;
+
+    if (!default_context_init_intern()
+            || !OPENSSL_atexit(do_default_context_deinit)
+            || !ossl_method_store_init()
+            || !OPENSSL_atexit(&ossl_method_store_cleanup)
+            /*
+             * TODO(3.0): Remove me. This is just a dummy call to demonstrate
+             * making EVP calls from within the FIPS module.
+             */
+            || !dummy_evp_call()) {
+        OPENSSL_cleanup();
+        return 0;
+    }
+
```

Comment:
This section is still problematic.  There should be no reason to do all the manual initialization here, since the affected stores (should) self-initialize on demand.

Also, I'm surprised to not see a teardown function calling `OPENSSL_cleanup`, I though that would be a fairly obvious thing to do.

Finally, I understand that you worry about multiple init of this module, due to the possibility of multiple library contexts.  I'm not sure how this helps, though (with the lack of a teardown that cleans up), but also, the experience with #8828 could suggest that an init counter would be useful, but here in `OSSL_provider_init` and associated teardown function rather than in the ex_data code.

Reply:
> This section is still problematic. There should be no reason to do all the manual initialization here, since the affected stores (should) self-initialize on demand.

We have no mechanism to enable that at the moment since RUN_ONCE just doesn't work in the general case. It could be useful where no actual resources are allocated that need to be later freed (such as where we just assign an ex_data index), but if we allocate resources we're a bit stuck. Even the ref counter doesn't work in that case since we cannot guarantee that if the ref counter goes to zero that the .bss and .data will be reinitialised before the provider is loaded again.

> Also, I'm surprised to not see a teardown function calling OPENSSL_cleanup, I though that would be a fairly obvious thing to do.

Since the only thing the teardown would need to do is call OPENSSL_cleanup, and the teardown function and OPENSSL_cleanup have the same signature, I just used OPENSSL_cleanup directly as the teardown function:

```
static const OSSL_DISPATCH fips_dispatch_table[] = {
    { OSSL_FUNC_PROVIDER_TEARDOWN, (void (*)(void))OPENSSL_cleanup },
```
I could insert an intermediate teardown, but it seems to add little value.

> Finally, I understand that you worry about multiple init of this module, due to the possibility of multiple library contexts. I'm not sure how this helps, though (with the lack of a teardown that cleans up), but also, the experience with #8828 could suggest that an init counter would be useful, but here in OSSL_provider_init and associated teardown function rather than in the ex_data code.

Yes there is still an issue here. As we discussed offline last week I think we need to have OSSL_provider_init return a provider side handle that corresponds to the libcrypto side OPENSSL_CTX object.

Reply:
[sigh]

The elepahnt in the room is that we're doing everything we can to avoid touching shared object init/fini type of constructors/destructors.  The more I see us doing all kinds of not very elegant stuff to avoid that, the more I'm wondering if we're not making a mistake.  This discussion is pretty indicative.

Reply:
From the "Self Testing" section of the design document:

"The module's Default Entry Point (DEP), the ".init" function in Linux libraries, will set a module variable (probably the state variable). This variable will be checked in the OSSL_provider_init() and, if set (which it always will be), will verify the values in the file. This two-step process meets the FIPS requirement of having the DEP ensure the tests are run, but allows us to implement the tests when the rest of the module is being initialized during normal operation."

So we have already recognised the need for a ".init" in the FIPS module.

Reply:
FIPS requires the .init constructor or equivalent (if available).  The note from the standard (whose name escapes me) even includes sample code for Linux/Unix and Windows.  A FIPS provider wouldn't be limited to not doing more in its constructor, so long as it doesn't offer any services before finishing its self tests.

Is it too much of a stretch going from _providers are shared libraries_ to _providers are shared libraries with initialisers_?

The alternative seems to be providers can be loaded but never unloaded and there might be restrictions on when they can be loaded.